### PR TITLE
Feature/transpose

### PR DESCRIPTION
**Pull request by kidder**

### CODE REVIEW

File: src/Numerical/LinearOperators/Transpose.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines function transpose
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Variables.hpp"
```
Comment (nilsdeppe) [inline src/Numerical/LinearOperators/Transpose.hpp]: This include isn't needed, nothing refers to Variables.

File: src/Numerical/LinearOperators/Transpose.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+/// \example
+/// \snippet Test_Transpose.cpp transpose_by_not_null_example
+/// \snippet Test_Transpose.cpp return_transpose_example
+/// \snippet Test_Transpose.cpp partial_transpose_example
+///
+/// \tparam U the type of data to be transposed
+/// \tparam T the type of the transposed data
+template <typename U, typename T>
+void transpose(const U& u, const size_t chunk_size,
+               const size_t number_of_chunks, gsl::not_null<T*> transpose) {
```
Comment (nilsdeppe) [inline src/Numerical/LinearOperators/Transpose.hpp]: take a `const` pointer to a non-`const` object

File: src/Numerical/LinearOperators/Transpose.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+#include "Utilities/Gsl.hpp"
+
+// @{
+/// \ingroup NumericalAlgorithms
+/// \brief Function to compute transposed data.
+///
+/// The primary use of this function is to rearrange the memory layout so that
+/// another function can operate on contiguous chunks of memory.
+///
+/// \requires `transpose.size()` to be the product of `number_of_chunks` and
```
Comment (nilsdeppe) [inline src/Numerical/LinearOperators/Transpose.hpp:21]: `u` has `.data()`

File: tests/Unit/Numerical/LinearOperators/Test_Transpose.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  DataVector transposed_data(n_pts, 0.);
+  transposed_data = transpose(data, chunk_size, number_of_chunks);
+  for (size_t i = 0; i < chunk_size; ++i) {
+    for (size_t j = 0; j < number_of_chunks; ++j) {
+      CHECK(data[i + chunk_size * j] ==
+            transposed_data[j + number_of_chunks * i]);
+    }
+  }
+  /// [return_transpose_example]
+  for (auto& p : transposed_data) {
```
Comment (nilsdeppe) [inline tests/Unit/Numerical/LinearOperators/Test_Transpose.cpp]: use `std::fill`

File: tests/Unit/Numerical/LinearOperators/Test_Transpose.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  DataVector ref_to_data;
+  ref_to_data.set_data_ref(&data);
+  DataVector ref_to_transposed_data;
+  ref_to_transposed_data.set_data_ref(&transposed_data);
+  CHECK(not ref_to_data.is_owning());
+  CHECK(not ref_to_transposed_data.is_owning());
+  ref_to_transposed_data = transpose(ref_to_data, chunk_size, number_of_chunks);
+  for (size_t i = 0; i < chunk_size; ++i) {
+    for (size_t j = 0; j < number_of_chunks; ++j) {
+      CHECK(ref_to_data[i + chunk_size * j] ==
```
Comment (nilsdeppe) [inline tests/Unit/Numerical/LinearOperators/Test_Transpose.cpp]: Should also check the referred to data

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Numerical/LinearOperators/Transpose.hpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+/// filling `transpose` (or the returned object) with the transpose of that
+/// matrix.
+///
+/// \example
+/// \snippet Test_Transpose.cpp transpose_by_not_null_example
+/// \snippet Test_Transpose.cpp return_transpose_example
+/// \snippet Test_Transpose.cpp partial_transpose_example
+///
+/// \tparam U the type of data to be transposed
+/// \tparam T the type of the transposed data
```
Comment (fmahebert) [inline src/Numerical/LinearOperators/Transpose.hpp:42]: ~~Why two template parameters? In the tests the input and output are of the same type. Will there be use cases with different types?~~ Edit: Nevermind, the partial transpose test uses different Variable types...

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
Squash please :)

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**